### PR TITLE
[Event Hubs Client] Track Two: First Preview (Test Infrastructure and Team Feedback)

### PR DESCRIPTION
**Pull request by jsquire**
# Summary

The intent of these changes is to incorporate feedback from review of the preview API design with the Event Hubs service team and to enable dynamic management of an Event Hubs namespace by the Live tests.

### Goals

- Rename the concepts of "senders" and "receivers" to "producers" and "consumers", covering all associated types and their documentation.

- Incorporate feedback for type names and mandate the event position be specified when creating a receiver. 

- Enable each live test to execute within the context of a dedicated Event Hub instance in order to ensure isolation to prevent tests from interacting with one another, and to enable tests to safely run in parallel.

- Restore the ability to send a single event from the `EventProducer`.

# Last Upstream Rebase

Wednesday, June 19, 2019  8:36am (EDT)

# Resources

- [.NET Event Hubs Client: Track Two Proposal (First Preview)](https://gist.github.com/jsquire/75b3a3090b6b4553aa8ceb798b6fc87d)
- [.NET Event Hubs Client: Deferred Features and Functionality](https://gist.github.com/jsquire/f88922faa0f457b503234c6f168e20c9)

### CODE REVIEW

File: eng/Packages.Data.props [hunk 1]
```diff
@@ -46,7 +47,7 @@
     <PackageReference Update="Microsoft.Extensions.Logging.Abstractions" Version="2.2.0" />
     <PackageReference Update="Microsoft.Extensions.Options" Version="2.2.0" />
     <PackageReference Update="Microsoft.Extensions.PlatformAbstractions" Version="1.1.0" />
-    <PackageReference Update="Microsoft.IdentityModel.Clients.ActiveDirectory" Version="4.5.1" />
+    <PackageReference Update="Microsoft.IdentityModel.Clients.ActiveDirectory" Version="[3.14.2, 4.5.1]" />
```
Comment (jsquire) [inline eng/Packages.Data.props]: This was needed to resolve conflicts between the Event Hubs management library and the Track One Event Hubs library, by way of dependencies.   Happy to consider another approach if someone has one; this was the leave invasive change that I was able to make.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneEventSender.cs [hunk 2]
```diff
@@ -60,7 +60,7 @@ public TrackOneEventSender(Func<TrackOne.EventDataSender> trackOneSenderFactory)
         /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
         ///
         public override Task SendAsync(IEnumerable<EventData> events,
-                                       EventBatchingOptions batchOptions,
+                                       SendOptions batchOptions,
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneEventSender.cs]: nit:  Fix naming.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Core/TransportEventSender.cs [hunk 3]
```diff
@@ -25,7 +25,7 @@ internal abstract class TransportEventSender
         /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
         ///
         public abstract Task SendAsync(IEnumerable<EventData> events,
-                                       EventBatchingOptions batchOptions,
+                                       SendOptions batchOptions,
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Core/TransportEventSender.cs]: nit: Fix naming

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventSender.cs [hunk 4]
```diff
@@ -205,5 +245,21 @@ protected EventSender()
         ///
         [EditorBrowsable(EditorBrowsableState.Never)]
         public override string ToString() => base.ToString();
+
+        /// <summary>
+        ///   Yeilds an <paramref name="eventData" /> as a single element
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/EventSender.cs]: nit:  Spelling

File: sdk/eventhub/Azure.Messaging.EventHubs/src/TrackOneClient/Azure.Messaging.EventHubs.TrackOne.csproj [hunk 5]
```diff
@@ -24,7 +24,7 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.Azure.Amqp" PrivateAssets="All" />
     <PackageReference Include="Microsoft.Azure.Services.AppAuthentication" PrivateAssets="All" />
-    <PackageReference Include="Microsoft.IdentityModel.Clients.ActiveDirectory" PrivateAssets="All" />
+    <PackageReference Include="Microsoft.IdentityModel.Clients.ActiveDirectory" PrivateAssets="All"/>
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/TrackOneClient/Azure.Messaging.EventHubs.TrackOne.csproj]: nit:  Unintentional change.   Restore.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/EventHubClient/EventHubClientLiveTests.cs [hunk 6]
```diff
@@ -140,8 +140,7 @@ await using (var client = new EventHubClient(connectionString, scope.EventHubNam
                     Assert.That(properties, Is.Not.Null, "A set of properties should have been returned.");
                     Assert.That(properties.Path, Is.EqualTo(scope.EventHubName), "The property Event Hub name should match the scope.");
                     Assert.That(properties.PartitionIds.Length, Is.EqualTo(partitionCount), "The properties should have the requested number of partitions.");
-                    Assert.That(properties.CreatedAtUtc, Is.LessThanOrEqualTo(DateTime.UtcNow), "The Event Hub should have been created prior to now.");
-                    Assert.That(properties.PropertyRetrievalTimeUtc, Is.EqualTo(DateTime.UtcNow).Within(TimeSpan.FromSeconds(30)), "The property retrieval time should denote the current time.");
+                    Assert.That(properties.CreatedAtUtc, Is.EqualTo(DateTime.UtcNow).Within(TimeSpan.FromSeconds(1)), "The Event Hub should have been created prior to now.");
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/tests/EventHubClient/EventHubClientLiveTests.cs]: nit:  Tweak message;  it's not entirely accurate any longer.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/EventHubScope.cs [hunk 7]
```diff
@@ -99,22 +141,159 @@ public ValueTask DisposeAsync()
...
+        /// <param name="exponentialBackoffSeconds">The exponential backoff amount,, in seconds.</param>
+        /// <param name="baseJitterSeconds">The amount of base jitter to include, in seconds.</param>
+        ///
+        /// <returns>The interval to wait before retrying the attempted operation.</returns>
+        ///
+        private static TimeSpan CalculateRetryDelay(int attempt, double exponentialBackoffSeconds, double baseJitterSeconds) =>
+            TimeSpan.FromSeconds((Math.Pow(2, attempt) * exponentialBackoffSeconds) + (RandomNumberGenerator.Value.NextDouble() * baseJitterSeconds));
+
+        /// <summary>
+        ///   Aquires an <see cref="AccessToken"/> for use with the Event Hubs management client.
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/EventHubScope.cs]: nit:  Fix type in <see>

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventSender.cs [hunk 8]
```diff
@@ -205,5 +245,21 @@ protected EventSender()
...
+        ///   enumerable, for use with the methods that operate on a batch
+        ///   of events.
+        /// </summary>
+        ///
+        /// <param name="eventData">The event to yield.</param>
+        ///
+        /// <returns>The requested event in the form of an enumerable.</returns>
+        ///
+        private IEnumerable<EventData> YieldEvent(EventData eventData)
```
Comment (tg-msft) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/EventSender.cs]: Nit - I think the name `AsEnumerable` is more common here, but I don't feel strongly as it's not public.

**Review by tg-msft — APPROVED** [review]
The changes look good.  I like the look of  the explicit `EventPosition.Latest` in the updated tests.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventSender.cs (hunk 8 above)
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventSender.cs]: I had spun on this a bit;  there was a bunch of opinions to both use `AsEnumerable` and not, though most of those were aimed at extension methods and confusion.   Given that this is private, can't hurt for clarity.   I'll tweak it.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneEventHubClient.cs [hunk 9]
```diff
@@ -206,101 +205,103 @@ public override async Task<EventHubProperties> GetPropertiesAsync(CancellationTo
...
-        /// <param name="senderOptions">The set of options to apply when creating the sender.</param>
+        /// <param name="producerOptions">The set of options to apply when creating the producer.</param>
         ///
-        /// <returns>An event sender configured in the requested manner.</returns>
+        /// <returns>An producer configured in the requested manner.</returns>
         ///
-        public override EventSender CreateSender(EventSenderOptions senderOptions)
+        public override EventHubProducer CreateProducer(EventHubProducerOptions producerOptions)
         {
             TrackOne.EventDataSender CreateSenderFactory()
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneEventHubClient.cs]: These should continue to be `Sender` nomenclature, as they refer to the track one types, which do not use the new nomenclature.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneEventHubClient.cs [hunk 10]
```diff
@@ -206,101 +205,103 @@ public override async Task<EventHubProperties> GetPropertiesAsync(CancellationTo
...
+        ///   Event Hub, grouped together in batches.  Depending on the <paramref name="producerOptions"/>
+        ///   specified, the producer may be created to allow event data to be automatically routed to an available
         ///   partition or specific to a partition.
         /// </summary>
         ///
-        /// <param name="senderOptions">The set of options to apply when creating the sender.</param>
+        /// <param name="producerOptions">The set of options to apply when creating the producer.</param>
         ///
-        /// <returns>An event sender configured in the requested manner.</returns>
+        /// <returns>An producer configured in the requested manner.</returns>
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneEventHubClient.cs]: nit:   "An producer"

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneEventHubClient.cs [hunk 11]
```diff
@@ -206,101 +205,103 @@ public override async Task<EventHubProperties> GetPropertiesAsync(CancellationTo
...
+                new TrackOneEventHubProducer(CreateSenderFactory),
                 TrackOneClient.EventHubName,
-                senderOptions
+                producerOptions
             );
         }
 
         /// <summary>
-        ///   Creates an event receiver responsible for reading <see cref="EventData" /> from a specific Event Hub partition,
+        ///   Creates a consumer responsible for reading <see cref="EventData" /> from a specific Event Hub partition,
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneEventHubClient.cs]: nit:  "An Event Hub consumer"

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneEventHubClient.cs [hunk 12]
```diff
@@ -206,101 +205,103 @@ public override async Task<EventHubProperties> GetPropertiesAsync(CancellationTo
...
-        /// <returns>An event receiver configured in the requested manner.</returns>
+        /// <returns>An consumer configured in the requested manner.</returns>
         ///
-        public override EventReceiver CreateReceiver(string partitionId,
-                                                     EventReceiverOptions receiverOptions)
+        public override EventHubConsumer CreateConsumer(string partitionId,
+                                                        EventPosition eventPosition,
+                                                        EventHubConsumerOptions consumerOptions)
         {
             TrackOne.PartitionReceiver CreateReceiverFactory()
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneEventHubClient.cs]: These should continue to be `Receiver` nomenclature, as they refer to the track one types, which do not use the new nomenclature.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Core/TransportEventHubClient.cs [hunk 13]
```diff
@@ -38,41 +38,43 @@ internal abstract class TransportEventHubClient
...
+        ///   By default, consumers are created as non-exclusive.
         /// </summary>
         ///
         /// <param name="partitionId">The identifier of the Event Hub partition from which events will be received.</param>
-        /// <param name="receiverOptions">The set of options to apply when creating the receiver.</param>
+        /// <param name="eventPosition">The position within the partition where the consumer should begin reading events.</param>
+        /// <param name="consumerOptions">The set of options to apply when creating the consumer.</param>
         ///
-        /// <returns>An event receiver configured in the requested manner.</returns>
+        /// <returns>An consumer configured in the requested manner.</returns>
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Core/TransportEventHubClient.cs]: nit:  "an consumer"

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Core/TransportEventHubClient.cs [hunk 14]
```diff
@@ -38,41 +38,43 @@ internal abstract class TransportEventHubClient
...
+        ///   Event Hub, grouped together in batches.  Depending on the <paramref name="producerOptions"/>
+        ///   specified, the producer may be created to allow event data to be automatically routed to an available
         ///   partition or specific to a partition.
         /// </summary>
         ///
-        /// <param name="senderOptions">The set of options to apply when creating the sender.</param>
+        /// <param name="producerOptions">The set of options to apply when creating the producer.</param>
         ///
-        /// <returns>An event sender configured in the requested manner.</returns>
+        /// <returns>An producer configured in the requested manner.</returns>
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Core/TransportEventHubClient.cs]: nit:  "an producer"

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneEventHubClient.cs [hunk 15]
```diff
@@ -206,101 +205,103 @@ public override async Task<EventHubProperties> GetPropertiesAsync(CancellationTo
...
+        ///   By default, consumers are created as non-exclusive.
         /// </summary>
         ///
         /// <param name="partitionId">The identifier of the Event Hub partition from which events will be received.</param>
-        /// <param name="receiverOptions">The set of options to apply when creating the receiver.</param>
+        /// <param name="eventPosition">The position within the partition where the consumer should begin reading events.</param>
+        /// <param name="consumerOptions">The set of options to apply when creating the consumer.</param>
         ///
-        /// <returns>An event receiver configured in the requested manner.</returns>
+        /// <returns>An consumer configured in the requested manner.</returns>
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneEventHubClient.cs]: nit:  "an consumer"

File: sdk/eventhub/Azure.Messaging.EventHubs/src/SendOptions.cs [hunk 16]
```diff
@@ -47,7 +47,7 @@ private int MaximumSizeInBytes
         ///   specified directly when sending the batch.
         /// </summary>
         ///
-        /// <value>The partition hashing key to associate with the event batch.</value>
+        /// <value>The partition hashing key to associate with the event or batch or events.</value>
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/SendOptions.cs]: nit:  Batch of events

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventSender.cs (hunk 8 above)
Comment (KrzysztofCwalina) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventSender.cs]: why not return an array? return new EventData[] {eventData};

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventSender.cs]: In retrospect,  I'm not really saving anything by the indirection.   I refactored to the array approach.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducerOptions.cs [hunk 17]
```diff
@@ -7,24 +7,24 @@
...
     {
         /// <summary>The timeout that will be used by default for sending events.</summary>
         private TimeSpan? _timeout = TimeSpan.FromMinutes(1);
 
         /// <summary>
-        ///   The identifier of the Event Hub partition that the <see cref="EventSender" /> will be bound to,
+        ///   The identifier of the Event Hub partition that the <see cref="EventHubProducer" /> will be bound to,
         ///   limiting it to sending events to only that partition.
         ///
         ///   If the identifier is not spedified, the Event Hubs service will be responsible for routing events that
```
Comment (conniey) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducerOptions.cs]: specified

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducerOptions.cs [hunk 18]
```diff
@@ -7,24 +7,24 @@
 namespace Azure.Messaging.EventHubs
```
Comment (conniey) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducerOptions.cs]: There is a spelling mistake, "considered"
https://github.com/jsquire/azure-sdk-for-net/blob/2568fe0c58b2117904e857c0faff34c8e440fe04/sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducerOptions.cs#L52

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducerOptions.cs]: Thank you!

**Pull request merged**